[config_mgmt.py]: Shutdown the interfaces after config validation.

praveen-li opened the pull request:
Changes:
-- Shutdown the interfaces after config validation while Dy Port Breakout.
-- lib.py: to avoid loop dependencies b/w config_mgmt.py and main.py
   moved few helper funtions in lib.py.

Signed-off-by: Praveen Chaudhary pchaudhary@linkedin.com

<!--
Please make sure you've read and understood our contributing guidelines:
https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

If this is a bug fix, make sure your description includes "closes #xxxx",
"fixes #xxxx" or "resolves #xxxx" so that GitHub automatically closes the related
issue when the PR is merged.

If you are adding/modifying/removing any command or utility script, please also
make sure to add/modify/remove any unit tests from the sonic-utilities-tests
directory as appropriate.

If you are modifying or removing an existing 'show', 'config' or 'sonic-clear'
subcommand, or you are adding a new subcommand, please make sure you also
update the Command Line Reference Guide (doc/Command-Reference.md) to reflect
your changes.

Please provide the following information:
-->

**- What I did**
Changes:
-- Shutdown the interfaces after config validation while Dy Port Breakout.
-- lib.py: to avoid loop dependencies b/w config_mgmt.py and main.py
   moved few helper funtions in lib.py.

**- How I did it**
-- move few helper functions to lib.py.
-- move shutdown part to config_mgmt.py

**- How to verify it**
```
svc-lnos-user@server10:/home/svc-lnos-user/praveen/sonic-buildimage/src/sonic-swss/tests$ cat test_port_dpb_system.py | grep -A 4 -B 4 Uncomment

        # Verify port is removed from ACL table
        self.dvs_acl.verify_acl_table_count(1)

        #TBD: Uncomment this, or explain why Ethernet0 is being added back to ACL table
        # Also, string "None" is being added as port to ACL port list after the breakout
        self.dvs_acl.verify_acl_group_num(0)

        # Verify child ports are created.

svc-lnos-user@server10:/home/svc-lnos-user/praveen/sonic-buildimage/src/sonic-swss/tests$ sudo pytest --dvsname=vs-jk test_port_dpb_system.py -k test_port_breakout_with_acl --junitxml=report.xml | tee out
============================= test session starts ==============================
platform linux2 -- Python 2.7.12, pytest-3.3.0, py-1.6.0, pluggy-0.6.0
rootdir: /home/svc-lnos-user/praveen/sonic-buildimage/src/sonic-swss/tests, inifile:
collected 2 items

test_port_dpb_system.py .                                                [100%]

 generated xml file: /home/svc-lnos-user/praveen/sonic-buildimage/src/sonic-swss/tests/report.xml
============================== 1 tests deselected ==============================
=================== 1 passed, 1 deselected in 92.36 seconds ====================
```

```
Jun  5 06:16:37.007136 fec26aa9f1be INFO #sonic_yang: Try to load Data in the tree
Jun  5 06:16:37.007722 fec26aa9f1be INFO #configMgmt: Data Validation successful
Jun  5 06:16:37.031857 fec26aa9f1be INFO #configMgmt: shutdown Interfaces: {'PORT': {'Ethernet2': {'admin_status': 'down'}, 'Ethernet3': {'admin_status': 'down'}, 'Ethernet0': {'admin_status': 'down'}, 'Ethernet1': {'admin_status': 'down'}}}
Jun  5 06:16:37.032485 fec26aa9f1be INFO #configMgmt: Write in DB: {'PORT': OrderedDict([('Ethernet0', {'admin_status': 'down'}), ('Ethernet1', {'admin_status': 'down'}), ('Ethernet2', {'admin_status': 'down'}), ('Ethernet3', {'admin_status': 'down'})])}
Jun  5 06:16:37.033823 fec26aa9f1be NOTICE #orchagent: :- doPortTask: Set port Ethernet0 admin status to down
Jun  5 06:16:37.034270 fec26aa9f1be INFO #configMgmt: Write in DB: {u'PORT': OrderedDict([(u'Ethernet0', None), (u'Ethernet1', None), (u'Ethernet2', None), (u'Ethernet3', None)])}
Jun  5 06:16:37.034383 fec26aa9f1be NOTICE #orchagent: :- doPortTask: Set port Ethernet1 admin status to down
Jun  5 06:16:37.034486 fec26aa9f1be NOTICE #orchagent: :- doPortTask: Set port Ethernet2 admin status to down
Jun  5 06:16:37.034872 fec26aa9f1be NOTICE #portsyncd: :- onMsg: nlmsg type:16 key:Ethernet0 admin:0 oper:0 addr:02:42:ac:11:00:0c ifindex:1199 master:0 type:tun
Jun  5 06:16:37.035004 fec26aa9f1be NOTICE #orchagent: :- doPortTask: Set port Ethernet3 admin status to down
Jun  5 06:16:37.035166 fec26aa9f1be NOTICE #portsyncd: :- onMsg: Publish Ethernet0(ok) to state db
Jun  5 06:16:37.035317 fec26aa9f1be INFO #configMgmt: Verify Port Deletion from Asic DB, Wait...
```

——Test 06/09/2020
```
Jun  9 23:58:26.590188 fec26aa9f1be INFO #configMgmt: shutdown Interfaces: {'PORT': {'Ethernet2': {'admin_status': 'down'}, 'Ethernet0': {'admin_status': 'down'}}}
Jun  9 23:58:26.590828 fec26aa9f1be INFO #configMgmt: Write in DB: {'PORT': OrderedDict([('Ethernet0', {'admin_status': 'down'}), ('Ethernet2', {'admin_status': 'down'})])}
Jun  9 23:58:26.591800 fec26aa9f1be INFO #configMgmt: Write in DB: {u'PORT': OrderedDict([(u'Ethernet0', None), (u'Ethernet2', None)])}
Jun  9 23:58:26.592121 fec26aa9f1be INFO #configMgmt: Verify Port Deletion from Asic DB, Wait...

svc-lnos-user@server10:/home/svc-lnos-user/praveen/sonic-buildimage/src/sonic-swss/tests$ cat test_port_dpb_system.py | grep -A 4 -B 4 Uncomment

        # Verify port is removed from ACL table
        self.dvs_acl.verify_acl_table_count(1)

        #TBD: Uncomment this, or explain why Ethernet0 is being added back to ACL table
        # Also, string "None" is being added as port to ACL port list after the breakout
        self.dvs_acl.verify_acl_group_num(0)

        # Verify child ports are created.
svc-lnos-user@server10:/home/svc-lnos-user/praveen/sonic-buildimage/src/sonic-swss/tests$ sudo pytest --dvsname=vs-jk test_port_dpb_system.py -k test_port_breakout_with_acl --junitxml=report.xml | tee out
============================= test session starts ==============================
platform linux2 -- Python 2.7.12, pytest-3.3.0, py-1.6.0, pluggy-0.6.0
rootdir: /home/svc-lnos-user/praveen/sonic-buildimage/src/sonic-swss/tests, inifile:
collected 2 items

test_port_dpb_system.py .                                                [100%]

 generated xml file: /home/svc-lnos-user/praveen/sonic-buildimage/src/sonic-swss/tests/report.xml
============================== 1 tests deselected ==============================
=================== 1 passed, 1 deselected in 93.69 seconds ====================
```
**- Previous command output (if the output of a command-line utility has changed)**

**- New command output (if the output of a command-line utility has changed)**

Commit:
[config_mgmt.py]: Shutdown the interfaces after config validation.

Changes:
-- Shutdown the interfaces after config validation while Dy Port Breakout.
-- lib.py: to avoid loop dependencies b/w config_mgmt.py and main.py
   moved few helper funtions in lib.py.

Signed-off-by: Praveen Chaudhary pchaudhary@linkedin.com

vasant17 commented on config/config_mgmt.py:
```diff
@@ -385,6 +389,23 @@ def validateConfigData(self):
+        """ shut down all the interfaces before deletion """
+        shutDownConf = dict(); shutDownConf["PORT"] = dict()
+        for intf in ports:
+            if not interface_name_is_valid(intf):
```
Why do we need to check this again here? When we receive delete and add ports to method breakOutPort(), aren't they already validated?

praveen-li replied:
yeah, except Parent port, nothing else is validated before.

vasant17 replied:
ok, then all of it should be validated in one place. I feel when something comes to this function, it should all be validated. Just like the way we validate CLI command options, breakout mode, root port name in pyclick, list of add ports and delete ports should also be verified at the same time! could you please explain me the reason for verifying add and delete port lists in the first place? Shouldn't we assume that they are always correct as they are being generated from platform.json? @samaity : Could you please comment on my assumption?

praveen-li replied:
This seems a good idea to validated all del_ports in PY Click itself instead of config_mgmt.py, because we should call the breakOut API only after this validation, I will move the code to main.py.
Validation is for configDb, we have port info from platform.json but we need to validate it in configDB as well that PORT really exists.

praveen-li replied:
Done

vasant17 commented on config/main.py:
```diff
@@ -1680,12 +1590,7 @@ def breakout(ctx, interface_name, mode, verbose, force_remove_dependencies, load
     del_intf_dict = {intf: del_ports[intf]["speed"] for intf in del_ports}
```
Seems duplicate line. Can we please delete it?

praveen-li replied:
@samaity : Double confirmation, I agree we can delete this line, Should I go ahead.

samaity replied:
Yes, kindly do it. I have already deleted the duplicate line in the upstream code.

vasant17 commented on config/config_mgmt.py:
```diff
@@ -232,9 +234,11 @@ def breakOutPort(self, delPorts=list(), addPorts= list(), portJson=dict(), \
+            # shutdown the ports, Update deletion of Config,
+            # verify in Asic DB for port deletion,
             # then update addition of ports in config DB.
+            self.shutdownIntf(delPorts)
```
Do we need to set admin_status of added ports to UP at line 247? If its done somewhere else, is that right place?

praveen-li replied:
It is decided by portJson dict passed to breakout_Ports APIs by Py Click. As I know it is always set to UP.

vasant17 replied:
Nope, I dont think so, could you please point me to the line where its set to UP?, below is the line where we create that json.
```
portJson = dict(); portJson['PORT'] = port_dict
```

I dont see where the interface added to port_dict has admin_status set to UP.

samaity replied:
as Praveen said, it's always set to up when we pass the dictionary.
[portconfig.py#L193](https://github.com/zhenggen-xu/sonic-buildimage/blob/sonic-cfg-mgmt/src/sonic-config-engine/portconfig.py#L193)

vasant17 replied:
OK, But I don't see uniformity here. When we call breakOutPort(), we expect it to shutdown ports, delete ports, and add ports. But we are deciding admin_status of newly created ports, outside of this function. OK, I will let Praveen make the call.

praveen-li replied:
Since new config for added port is controlled by platform.json and not by configdb.json, we have to pass lanes, speed, index, admin_staus from py click. There is uniformity because PY click just passes it but do not set it to down or up. We set it down and up after config validation in config_mgmt.py only. I feel, this is good,

Commit:
[config_mgmt.py]: Validate del ports before calling breakOutPorts API.

Changes:
-- Validate del ports before calling breakOutPorts API.
-- remove lib.py, since it is not needed now.

Signed-off-by: Praveen Chaudhary pchaudhary@linkedin.com

praveen-li commented:
@vasant17  @samaity  : in 2nd commit I am also removing new file lib.py [was added in 1st commit], since there is no need for it now. We call interface_name_is_valid() only in main.py now. 

So diff looks big, but it is just the move of some functions back to main.py from lib.py.
